chain/chain/src/store.rs

```diff
@@ -2535,61 +2642,476 @@ mod tests {
+            do_fork(
+                source_block.clone(),
+                state_root,
+                trie.clone(),
+                &mut chain,
+                simple_chain.length,
+                &mut blocks,
+                max_changes,
+                verbose,
+            );
+        }
 
-        assert!(chain.get_block(&blocks[4].hash()).is_ok());
-        assert!(chain.get_block(&blocks[5].hash()).is_err());
-        assert!(chain.get_block(&blocks[6].hash()).is_ok());
-        // block header should be available
-        assert!(chain.get_block_header(&blocks[5].hash()).is_ok());
-        assert!(chain.mut_store().get_all_block_hashes_by_height(5).is_err());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[4].hash()).is_ok());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[5].hash()).is_err());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[6].hash()).is_ok());
         assert!(check_refcount_map(&mut chain).is_ok());
+        // GC execution
+        let clear_data = chain.clear_data(trie.clone());
+        if clear_data.is_err() {
+            println!("clear data failed = {:?}", clear_data);
+            assert!(false);
+        }
+        assert!(check_refcount_map(&mut chain).is_ok());
+
+        // Validation phase
+        let mut chain2 = get_chain();
+        chain2.epoch_length = 10;
+        let trie2 = chain2.runtime_adapter.get_trie().clone();
+
+        // Find gc_height
+        let mut gc_height = chains[0].length - 51;
+        for (i, simple_chain) in chains.iter().enumerate() {
+            if (!simple_chain.is_removed) && gc_height > simple_chain.from && i > 0 {
+                gc_height = simple_chain.from
+            }
+        }
+        if verbose {
+            println!("GC height = {:?}", gc_height);
+        }
+
+        let mut start_index = 1; // zero is for genesis
+        let mut state_roots = vec![Trie::empty_root()];
+        let mut trie_data = HashSet::new();
+
+        for simple_chain in chains {
+            if simple_chain.is_removed {
+                start_index += simple_chain.length;
+                continue;
+            }
+            let (_, mut state_root, _) = blocks[simple_chain.from as usize].clone();
+
+            if trie.iter(&state_root).is_ok() {
+                let a = trie
+                    .iter(&state_root)
+                    .unwrap()
+                    .map(|item| {
+                        if item.is_ok() {
+                            let (key, _) = item.unwrap();
+                            key
+                        } else {
+                            vec![]
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                trie_data.insert(a);
+            }
+
+            for i in start_index..start_index + simple_chain.length {
+                let mut store_update = chain2.mut_store().store_update();
```

Comment:
what's the difference between the way we apply changes to chain and chain2?

Reply:
Test infrastructure do the following:
1. Build Chain 1 from the full data, then GC it.
2. Build Chain 2 from the data after removing everything that should be removed after GC.
3. Make sure that Chain 1 == Chain 2.

Reply:
Makes sense

Reply:
I guess, this explanation is valuable enough to become a comment in the code